domain/estimators/recurrent_logistic.py

```diff
@@ -0,0 +1,378 @@
+        """
+        self._domain_idx_df = pd.DataFrame(domain_idx_df)
+
+    def _gen_train_tensor(self, row, attr, values):
+        """
+        Returns featurized tensor using training corpus (statistics are
+        based on current best predicted values).
+
+        :param row: (namedtuple, recarray, dict) current values
+        :param attr: (str) attribute for :param values:
+        :param values: (list[str]) values to generate features for
+        """
+        return self._gen_feat_tensor(row, attr, values, self.train_featurizers)
+
+    def _gen_test_tensor(self, row, attr, values):
+        """
+        Returns featurized tensor using initial corpus (statistics are
+        based on initial values).
+
+        :param row: (namedtuple, recarray, dict) current values
+        :param attr: (str) attribute for :param values:
+        :param values: (list[str]) values to generate features for
+        """
+        return self._gen_feat_tensor(row, attr, values, self.test_featurizers)
+
+    def _gen_feat_tensor(self, row, attr, values, featurizers):
+        """
+        :param row: (namedtuple, recarray, dict) current values
+        :param attr: (str) attribute for :param values:
+        :param values: (list[str]) values to generate features for
+        :param featurizers: (list[Featurizer]) featurizers to use
+        """
+        # Fastpath for 1 featurizer
+        if len(featurizers) == 1:
+            return featurizers[0].create_tensor(row, attr, values)
+
+        out_tensor = torch.zeros(len(values), sum(feat.num_features() for feat in featurizers))
+        # iterate through each featurizer
+        feat_idx = 0
+        for featurizer in featurizers:
+            feat_tensor = featurizer.create_tensor(row, attr, values)
+            assert(feat_tensor.shape[0] == len(values))
+            assert(feat_tensor.shape[1] == featurizer.num_features())
+
+            out_tensor[:, feat_idx:feat_idx+featurizer.num_features()] = feat_tensor
+            feat_idx += featurizer.num_features()
+
+        return out_tensor
+
+    def forward(self, X):
+        linear = X.matmul(self._W) + self._B
+        return torch.sigmoid(linear)
+
+    def train(self, num_recur=1, num_epochs=3, batch_size=32):
+        """
+        Trains the LR model. Updates the current values with the maximum a posteriori after
+        each recurrent iteration.
+        :param num_recur: (int) number of times to train model.
+        :param num_epochs: (int) number of epochs PER recurrent iteration.
+        :param batch_size: (int) size of batch in stochastic gradient descent.
+        """
+        batch_losses = []
+        for recur_idx in range(1, num_recur + 1):
+            # We need to update our statistics in featurizers on the 2nd and
+            # later iteration with our newest current values.
+            if recur_idx > 1:
+                self._update_featurizers()
+            self._update_training_data()
```

Comment:
Why is it necessary to update training data after each iteration? Does training mutate the training data?